# 01MIAR - Actividad Whitepapers

## Introducción:

En el ámbito del desarrollo software cuando se intenta implementar una solución software no solo hay que prestar atención a llegar a una resolución concreta sino también a aplicar buenos patrones de desarrollo de código que permitan que sea más eficiente. En este documento se analiza y reflexiona sobre un artículo científico que aborda las principales malas prácticas que se llevan a cabo en el desarrollo de software en Python tanto para proyectos asociados al Machine Learning como para proyectos que no lo están.

Posteriormente, se evaluará cómo el hecho de aplicar tanto los patrones de diseño ineficientes como los eficientes influyen en el coste temporal de la ejecución del código Python. 

## Objetivos:

- Realizar un análisis reflexivo sobre el artículo científico y los resultados obtenidos.
- Servir como guía para demostrar de manera empírica cómo utilizar ciertos patrones de desarrollo software mejora la eficiencia del código.
- Ayudar a los desarrolladores a identificar y evitar el uso de prácticas poco recomendadas de desarrollo de software en Python.

## Artículo científico fuente:

El artículo científico sobre el que se basa este trabajo es el siguiente:

- [François Belias, Leuson Da Silva, Foutse Khomh, Cyrine Zid (2025), Performance Smells in ML and Non-ML Python Projects: A Comparative Study]
 [https://arxiv.org/abs/2504.20224](https://arxiv.org/abs/2504.20224)

La motivación de haber elegido este artículo científico es que aborda un tema muy relacionado con la asignatura, pudiendo servir como soporte y guía para principiantes en el mundo de la programación con Python, aprendiendo desde un primer momento a utilizar técnicas de desarrollo eficientes y sofisticadas como la comprensión de listas, diccionarios o conjuntos, así como asignaciones múltiples eficientes o comparaciones condicionales óptimas. 

Otro motivo de especial relevancia es que no solo se estudian los patrones incorrectos en código Python sino su distribución en proyectos de Machine learning frente a los que no lo son, distinguiendo incluso dentro de proyectos de Machine Learning en qué fases son más frecuentes.

## Resumen y reflexión del artículo científico

El artículo científico anteriormente citado analiza el impacto de las malas prácticas de rendimiento (performance smells) en código Python, especialmente comparando proyectos de Machine Learning (ML) con aquellos que no lo son (non-ML). A continuación, se plantean las tres principales preguntas a las que responde el estudio, sus respectivas respuestas y algunas conclusiones propias a partir de estas respuestas.

### RQ1 ¿ Como se distribuyen las malas prácticas en código python en proyectos de machine Learning frente a proyectos que no lo son?

La conclusión a la que llega el estudio es que en los proyectos de ML hay una mayor proporción de ineficiencias frente a los non-ML.. Desde mi punto de vista considero que tiene bastante sentido dado que en promedio los proyectos de ML suelen ser mucho más complejos, implican procesamiento de mayor volumen de datos y técnicas de programación más avanzadas

### RQ2 ¿Qué diferentes tipos de implementación equivocada podemos encontrarnos en proyectos de Machine Learning?

El artículo identifica un total de nueve performance smells en código Python. Estas malas prácticas afectan al rendimiento, la legibilidad y la eficiencia del software, y son fácilmente evitables mediante el uso de patrones idiomáticos del lenguaje. A continuación, se resumen las principales:

| Concepto                     | Mala práctica                                             | Buena práctica                                                  |
|-----------------------------|------------------------------------------------------------|------------------------------------------------------------------|
| List Comprehension          | Usar `for` con `append()` para construir listas            | Usar comprensiones de listas (`[x for x in ...]`)                |
| Set Comprehension           | Usar `for` con `add()` para construir sets                 | Usar comprensiones de sets (`{x for x in ...}`)                  |
| Dictionary Comprehension    | Usar `for` para asignar claves/valores a un diccionario    | Usar comprensiones de diccionarios (`{k: v for k, v in ...}`)    |
| Chain Compare               | Comparaciones separadas con `and`                         | Comparaciones encadenadas (`a < b <= c`)                         |
| Truth Value Test            | Comparar explícitamente con valores booleanos             | Usar la evaluación implícita de verdad (`if x:` en lugar de `if x != 0:`) |
| For-Else                    | Uso de bandera (`flag`) para saber si se salió del `for`   | Usar directamente el bloque `else` del `for`                     |
| Assign Multi Targets        | Asignaciones secuenciales una por una                     | Asignación múltiple (`a, b = b, a`)                              |
| Star-in-Func-Call           | Llamar funciones con múltiples argumentos explícitos       | Usar `*args` para desempaquetar listas o rangos                  |
| For-Multi Targets           | Iterar y luego extraer elementos manualmente (`x[0], x[1]`) | Desempaquetar directamente en el encabezado del bucle `for`     |

### RQ3: ¿Comó estan distribuidos estas malas prácticas de desarrollo a lo largo de las diferentes fases de un proyecto de machine Learning?

Se concluye que la fase de pre-procesamiento de datos es la fase de un proyecto de machine learning más susceptible de errores, acumulando la mayor parte de errores
en proyectos de aprendizaje profundo. Desde mi perspectiva tiene bastante sentido dado que la fase de pre-procesamiento de datos es la que menos limitada está y la que 
más depende del factor humano. Es decir, una vez se elije un modelo de machine learning está muy estandarizado y acotado como debe ser usado este modelo y por ello
siguiendo una metodología se es menos propenso a errores.

Sin embargo, en la fase de procesamiento de datos, el factor humano tiene más relevancia y hay más libertad de desarrollo sin seguir una metodología específica por lo que 
la probabilidad de producir errores es más alta. 


## Analisis del coste temporal empírico

Como complemento a lo aportado en el artículo científico el objetivo de este documento es enriquecer el conocimiento aportado haciendo un análisis del coste temporal empírico de aplicar unos determinados patrones de desarrollo de software frente a otros. Si llegamos a las mismas conclusiones de que a pequeña escala es más eficiente desarrollar de una manera que de otra, podremos concluir que esto en procesamientos que implican dimensiones mucho mayores del tamaño del problema a resolver implicarían costes elevadísimos tanto a diferentes niveles (computacional, temporal, energético, etc.)

La manera en la que se ha decidido medir el tiempo empírico ha sido creando una función personalizada y reutilizable. Si bien otra alternativa hubiera sido utilizar funciones mágicas de Jupyter Notebook, se ha optado por usar herramientas más nativas de Python para tener más flexibilidad a la hora de decidir el número de ejecuciones y el formato en el que se imprime la información. Además de medir el tiempo de está manera, dado que estamos evaluando operaciones con coste temporales muy pequeños, se ha optado por ejecutar la operación 100 veces para obtener un caso promedio y así tener información más fiable.

In [3]:
import time


def control_time_execution(function, n = 100):
    """ Imprime el tiempo de ejecución promedio de una función dada, ejecutándola n veces

    Keyword arguments:
    function -- función a medir
    n -- número de iteraciones (default 100)
    """
    acc = 0
    for _ in range(n):
        start = time.perf_counter() * 1000000
        function()
        end = time.perf_counter() * 1000000
        acc += ((end -start))
    mean_time = acc / n
    print(f"[{function.__name__}] Tiempo de ejecución promedio en {n} iteraciones: {(mean_time):.2f} microsegundos")
    return mean_time

def compare_two_functions(function_A, function_B, n = 100):
    """ Compara el tiempo de ejecución de dos funciones y muestra cuál es más rápida

    Keyword arguments:
    function_A -- primera función a comparar
    function_B -- segunda función a comparar
    n -- número de iteraciones (default 100)
    """
    t1 = control_time_execution(function_A, n)
    t2 = control_time_execution(function_B, n)
    if t1 > t2:
        porcentaje = (t1 - t2) / t1 * 100
        print(f'La función {function_B.__name__} es más rápida que {function_A.__name__}, con una disminución del tiempo de ejecución del {porcentaje:.2f}%.')
    else:
        porcentaje = (t2 - t1) / t2 * 100
        print(f'La función {function_A.__name__} es más rápida que {function_B.__name__}, con una disminución del tiempo de ejecución del {porcentaje:.2f}%.')


## ¿Qué malas prácticas se van a examinar? 

Se han seleccionado un subconjunto representativo de los "code smells", concretamente:

- Comprensiones de listas y diccionarios.
- Intercambio de variables : Este "code smell" es el que más se ha encontrado en los proyectos de machine Learning.
- Condiciones explícitas en contraste con implicitas.

### Comprensiones

Las comprensiones son una forma concisa, elegante y más eficiente de crear estructuras de datos como listas, conjuntos y diccionarios usando una sola línea de código, en lugar de usar bucles tradicionales. En el siguiente fragmento de código vamos a crear un listado, conjunto y diccionario de 100.000 elementos contrastando el coste temporal empiríco de usar Comprensiones frente a métodos iterativos más tradicionales. 

#### Listas

In [6]:
def iterative_list_generator():
    iterative_list = []
    for el in range(100000):
        iterative_list.append(el)
    return iterative_list

def comprehension_list_generator():
   return [el for el in range(100000)]

compare_two_functions(iterative_list_generator,  comprehension_list_generator)


[iterative_list_generator] Tiempo de ejecución promedio en 100 iteraciones: 4072.68 microsegundos
[comprehension_list_generator] Tiempo de ejecución promedio en 100 iteraciones: 2342.32 microsegundos
La función comprehension_list_generator es más rápida que iterative_list_generator, con una disminución del tiempo de ejecución del 42.49%.


Podemos reforzar está idea si analizamos el resumen del artículo "Python Functional Programming: Study of List Comprehensions and Lambda Functions Performance and Change-Proneness Risk" [1] que dice que en general el uso de listas de comprensión es mejor que usar bucles for.

#### Diccionarios

In [4]:
def iterative_dict_generator():
    iterative_dict = {}
    for el in range(100000):
        iterative_dict[el] = el
    return iterative_dict

def comprehension_dict_generator():
    return {el: el for el in range(100000)}

compare_two_functions(iterative_dict_generator, comprehension_dict_generator)

[iterative_dict_generator] Tiempo de ejecución promedio en 100 iteraciones: 6183.96 microsegundos
[comprehension_dict_generator] Tiempo de ejecución promedio en 100 iteraciones: 5862.59 microsegundos
La función comprehension_dict_generator es más rápida que iterative_dict_generator, con una disminución del tiempo de ejecución del 5.20%.


#### Asignación de múltiples valores

En la implementación de código es muy común encontrarse con el uso de variables auxiliares para realizar intercambios de valores entre variables. Python ofrece una manera más eficiente de desarrollar esta operación. En la función "multiple_assignments" veremos una manera más rudimentaria de implementar este intercambio de valores; sin embargo, en "multiple_assignments_optimized" vemos que el intercambio es más compacto.


In [14]:
def multiple_assignments():
    lst = [1, 2]
    for _ in range(10000):
        aux = lst[0]
        lst[0] = lst[1]
        lst[1] = aux
    return lst

def multiple_assignments_optimized():
    lst = [1, 2]
    for _ in range(10000):
        lst[0], lst[1] = lst[1], lst[0]
    return lst

compare_two_functions(multiple_assignments, multiple_assignments_optimized, 1000)

[multiple_assignments] Tiempo de ejecución promedio en 1000 iteraciones: 553.01 microsegundos
[multiple_assignments_optimized] Tiempo de ejecución promedio en 1000 iteraciones: 505.93 microsegundos
La función multiple_assignments_optimized es más rápida que multiple_assignments, con una disminución del tiempo de ejecución del 8.51%.


## Revisión explicita de condiciones

En este caso, la función "explicit_truth_value_test" hace una revisión más explícita de una condición, mientras que "implicit_truth_value_test" hace una revisión más implícita, aplicando con ello mejores prácticas de Python.

In [20]:
def explicit_truth_value_test():
    a = ''
    for n in range(100000):
        if n % 2 != 0:
            a = 'even'
        else:
            a = 'odd'
    return a

def implicit_truth_value_test():
    a = ''
    for n in range(100000):
        if n % 2:
            a = 'even'
        else:
            a = 'odd'

compare_two_functions(explicit_truth_value_test, implicit_truth_value_test)

[explicit_truth_value_test] Tiempo de ejecución promedio en 100 iteraciones: 3377.84 microsegundos
[implicit_truth_value_test] Tiempo de ejecución promedio en 100 iteraciones: 2735.67 microsegundos
La función implicit_truth_value_test es más rápida que explicit_truth_value_test, con una disminución del tiempo de ejecución del 19.01%.


## Conclusiones

En la siguiente tabla podemos apreciar los tiempos empíricos obtenidos para un determinada ejecución de este notebook, poniendo énfasis en que estos tiempos pueden cambiar dependiendo del hardware, plataforma, sistema operativo, etc., donde se ejecute este notebook.

| Prueba                              | Método tradicional (µs) | Método eficiente (µs) | Mejora (%) |
|--------------------------------------|-------------------------|-----------------------|------------|
| Generación de listas                 | 4135.97                 | 2191.18               | 47.02      |
| Generación de diccionarios           | 6183.96                 | 5862.59               | 5.20       |
| Asignación múltiple                  | 553.01                  | 505.93                | 8.51       |
| Truth Value Test (explícito vs impl.)| 3377.84                 | 2735.67               | 19.01      |

podemos llegar a las siguiente conclusiones:

- En general tal y como afirma el artículo cientifíco cuando implementamos software que sigue mejores prácticas Python hay una mejora del rendimiento y la velocidad empírica de ejecución.
- Se ha apreciado que en el caso de comprensiones de listas la mejora es considerablemente alta, llegando a un 45-60% de mejora entre diferentes ejecuciones de este notebook, cabe subrayar que 
estamos hablando de Python nativo en este artículo y que probablemente si lo contrastamos con Numpy, esté último dará aún mejores rendimientos, dado que trabaja de manera mucho más optimizada
listados de números en forma de numpy array. 

- La asignación múltiple es el fallo más común encontrado en los proyectos software de machine learning, si bien es el más abundante, no es el que más impacto tiene en el rendimiento, dado que su mejora
oscila entre el 1% y el 9% entre diferentes ejecuciones de este mismo jupyternotebook, pero podemos concluir que aunque sea una mejora muy baja, cuando lo extrapolamos a escenarios con mucho más datos
esta mejora se apreciará mucho más.

En definitiva, la adopción de buenas prácticas de desarrollo en Python no solo contribuye a la eficiencia y legibilidad del código, sino que también puede suponer una diferencia significativa en proyectos reales donde el volumen de datos y la complejidad de los procesos son mucho mayores. Implementar estos patrones desde las primeras fases del desarrollo permite construir soluciones más robustas, escalables y sostenibles, optimizando recursos y facilitando el mantenimiento a largo plazo. Por tanto, invertir en la calidad del código es invertir en el éxito y la competitividad de los proyectos de inteligencia artificial y ciencia de datos.

## Referencias

- [1] [ François Philippe Ossim Belias (2022), Python Functional Programming: Study of List Comprehensions and Lambda Functions Performance and Change-Proneness Risk]
 [https://publications.polymtl.ca/10764/](https://publications.polymtl.ca/10764/)